File: array-hash/347-Top-K-Frequent-Elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = defaultdict(int)
        frequency = [[] for _ in range(len(nums) + 1)]
        
        for n in nums:
            count[n] += 1
        ans = []
        for key, value in count.items():
            frequency[value].append(key)
        for i in range(len(frequency) - 1, -1, -1):
            for n in frequency[i]:
                ans.append(n)
                if len(ans) == k:
                    return ans
        return ans

File: reference/array-hash/347-Top-K-Frequent-Elements.go
func topKFrequent(nums []int, k int) []int {
	count := make(map[int]int)
	for _, num := range nums {
		count[num]++
	}

	n := len(nums)
	frequency := make([][]int, n+1)

	for num, freq := range count {
		frequency[freq] = append(frequency[freq], num)
	}

	var res []int
	for i := n; i > 0; i-- {
		for _, num := range frequency[i] {
			res = append(res, num)
			if len(res) == k {
				return res
			}
		}
	}

	return res
}